rust/loqui_client/src/client.rs

```diff
@@ -1,37 +1,37 @@
 use crate::connection_handler::{ConnectionHandler, InternalEvent};
+use crate::waiter::ResponseWaiter;
 use crate::Config;
 use failure::Error;
-use futures::sync::oneshot;
-use loqui_connection::{Encoder, LoquiError, Supervisor as SupervisedConnection};
+use loqui_connection::{Encoder, Supervisor as SupervisedConnection};
 use std::net::SocketAddr;
 use std::sync::Arc;
 use tokio::await;
 
 #[derive(Clone)]
 pub struct Client<E: Encoder> {
     connection: Arc<SupervisedConnection<ConnectionHandler<E>>>,
+    config: Arc<Config<E>>,
 }
 
 impl<E: Encoder> Client<E> {
     pub async fn connect(address: SocketAddr, config: Config<E>) -> Result<Client<E>, Error> {
         let config = Arc::new(config);
-        let handler_creator = move || ConnectionHandler::new(config.clone());
+        let handler_config = config.clone();
+        let handler_creator = move || ConnectionHandler::new(handler_config.clone());
         let connection = await!(SupervisedConnection::connect(address, handler_creator))?;
         let client = Self {
             connection: Arc::new(connection),
+            config,
         };
         Ok(client)
     }
 
     /// Send a request to the server.
     pub async fn request(&self, payload: E::Encoded) -> Result<E::Decoded, Error> {
```

Comment:
Q: Can we drop the `async` here, and just make this return the `awaitable` directly as an impl future?

Reply:
I'd prefer to keep it `async` so that it's a `std::future::Future`. I'd prefer not to return non-std futures in the client API.